### PR TITLE
Add: simple task pool without queue

### PR DESCRIPTION
**Pull request by watermelo**
**What this PR does**:
Simplify the task pool

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: sync/task_pool.go [hunk 1]
```diff
@@ -19,43 +19,224 @@ package gxsync
...
+
+	idx    uint32 // round robin index
+	qArray []chan task
+	wg     sync.WaitGroup
+
+	once sync.Once
+	done chan struct{}
+}
+
+// build a task pool
```
Comment (AlexStocks) [inline sync/task_pool.go:67]: the comment should begin with its func name

File: sync/options.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	if o.tQNumber > o.tQPoolSize {
+		o.tQNumber = o.tQPoolSize
+	}
+}
+
+type TaskPoolOption func(*TaskPoolOptions)
+
+// WithTaskPoolTaskQueueLength set @size of the task queue pool size
```
[the remark this answers is not in the document]
Comment (AlexStocks) [reply sync/options.go:43]: 如机器人提示，注释里函数名写错了

File: sync/task_pool.go [hunk 3]
```diff
@@ -32,69 +33,26 @@ import (
...
-	}
+// GenericTaskPool represents an generic task pool.
+type GenericTaskPool interface {
+	// AddTask wait idle worker add task
+	AddTask(t task) bool
+	// AddTaskAlways add task to queues or do it immediately
+	AddTaskAlways(t task)
+	// AddTaskBalance add task to idle queue
+	AddTaskBalance(t task)
+	Close()
```
Comment (AlexStocks) [inline sync/task_pool.go:46]: pls add comment for this func

**Review by AlexStocks — APPROVED** [review]

**Review by cityiron — APPROVED** [review]

File: sync/task_pool.go [hunk 4]
```diff
@@ -267,3 +221,66 @@ func (p *TaskPool) Close() {
...
+		}
+		<-p.sem
+	}()
+	for {
+		t()
+		t = <-p.work
+	}
+}
+
+func (p *taskPoolSimple) Close() {}
```
Comment (wongoo) [inline sync/task_pool.go:282]: when to use the simple task pool? why not implement the Close() func?

File: sync/task_pool_test.go [hunk 5]
```diff
@@ -252,21 +404,33 @@ func BenchmarkTaskPool_RandomTask(b *testing.B) {
...
-BenchmarkTaskPool_CPUTask/AddTask-8           	   72525	     18160 ns/op	       0 B/op	       0 allocs/op
-BenchmarkTaskPool_CPUTask/AddTaskAlways-8     	  606813	     16464 ns/op	      40 B/op	       0 allocs/op
-BenchmarkTaskPool_CPUTask/AddTaskBalance-8    	  137926	     17646 ns/op	       0 B/op	       0 allocs/op
-BenchmarkTaskPool_IOTask/AddTask-8            	   10000	    108520 ns/op	       0 B/op	       0 allocs/op
-BenchmarkTaskPool_IOTask/AddTaskAlways-8      	 1000000	      1236 ns/op	      95 B/op	       1 allocs/op
-BenchmarkTaskPool_IOTask/AddTaskBalance-8     	 1518144	       673 ns/op	      63 B/op	       0 allocs/op
-BenchmarkTaskPool_RandomTask/AddTask-8        	  497055	      2517 ns/op	       6 B/op	       0 allocs/op
-BenchmarkTaskPool_RandomTask/AddTaskAlways-8  	 2511391	       415 ns/op	      21 B/op	       0 allocs/op
-BenchmarkTaskPool_RandomTask/AddTaskBalance-8 	 1381711	       868 ns/op	      17 B/op	       0 allocs/op
+BenchmarkTaskPoolSimple_CountTask/AddTask-8              1693192               700 ns/op               0 B/op          0 allocs/op
```
Comment (wongoo) [inline sync/task_pool_test.go:410]: I think there should be a document to tell user when to use the TaskPool , and when to use the SimpleTaskTool?

Comment (AlexStocks) [reply sync/task_pool_test.go:410]: agree with u

File: sync/task_pool.go (hunk 4 above)
Comment (watermelo) [reply sync/task_pool.go]: done

File: sync/task_pool_test.go (hunk 5 above)
Comment (watermelo) [reply sync/task_pool_test.go:489]: The data as follows, run 100000 tasks and with 1000 size of taskpool.
```golang
// IOFunction
func newIOTask() (func(), *int64) {
	var cnt int64
	return func() {
		time.Sleep(700 * time.Microsecond)
	}, &cnt
}

=== RUN   TestWithoutPool
--- PASS: TestWithoutPool (0.04s)
    task_pool_test.go:384: Before Alloc = 0 MiB
    task_pool_test.go:385:      TotalAlloc = 0 MiB
    task_pool_test.go:386:      Sys = 66 MiB
    task_pool_test.go:387:      NumGC = 0
    task_pool_test.go:416: TestWithoutPool took 3460 goroutines
    task_pool_test.go:384: After Alloc = 2 MiB
    task_pool_test.go:385:      TotalAlloc = 8 MiB
    task_pool_test.go:386:      Sys = 69 MiB
    task_pool_test.go:387:      NumGC = 3
    task_pool_test.go:393: 
                 TestWithoutPool took 36.100024ms
PASS

=== RUN   TestWithSimplePool
--- PASS: TestWithSimplePool (0.13s)
    task_pool_test.go:385: Before Alloc = 0 MiB
    task_pool_test.go:386:      TotalAlloc = 0 MiB
    task_pool_test.go:387:      Sys = 68 MiB
    task_pool_test.go:388:      NumGC = 0
    task_pool_test.go:445: TestWithSimplePool took 1000 goroutines
    task_pool_test.go:385: After Alloc = 0 MiB
    task_pool_test.go:386:      TotalAlloc = 0 MiB
    task_pool_test.go:387:      Sys = 68 MiB
    task_pool_test.go:388:      NumGC = 0
    task_pool_test.go:394: 
                 TestWithSimplePool took 129.873271ms
PASS

=== RUN   TestWithSimplePoolUseAlways
--- PASS: TestWithSimplePoolUseAlways (0.06s)
    task_pool_test.go:385: Before Alloc = 0 MiB
    task_pool_test.go:386:      TotalAlloc = 0 MiB
    task_pool_test.go:387:      Sys = 66 MiB
    task_pool_test.go:388:      NumGC = 0
    task_pool_test.go:431: TestWithSimplePool took 2048 goroutines
    task_pool_test.go:385: After Alloc = 2 MiB
    task_pool_test.go:386:      TotalAlloc = 4 MiB
    task_pool_test.go:387:      Sys = 69 MiB
    task_pool_test.go:388:      NumGC = 1
    task_pool_test.go:394: 
                 TestWithSimplePool took 60.207644ms
PASS

=== RUN   TestWithPool
--- PASS: TestWithPool (0.14s)
    task_pool_test.go:384: Before Alloc = 0 MiB
    task_pool_test.go:385:      TotalAlloc = 0 MiB
    task_pool_test.go:386:      Sys = 66 MiB
    task_pool_test.go:387:      NumGC = 0
    task_pool_test.go:462: TestWithPool took 1000 goroutines
    task_pool_test.go:384: After Alloc = 0 MiB
    task_pool_test.go:385:      TotalAlloc = 0 MiB
    task_pool_test.go:386:      Sys = 68 MiB
    task_pool_test.go:387:      NumGC = 0
    task_pool_test.go:393: 
                 TestWithPool took 135.168986ms
PASS

=== RUN   TestWithPoolUseAlways
--- PASS: TestWithPoolUseAlways (0.07s)
    task_pool_test.go:384: Before Alloc = 0 MiB
    task_pool_test.go:385:      TotalAlloc = 0 MiB
    task_pool_test.go:386:      Sys = 66 MiB
    task_pool_test.go:387:      NumGC = 0
    task_pool_test.go:480: TestWithPoolUseAlways took 3280 goroutines
    task_pool_test.go:384: After Alloc = 2 MiB
    task_pool_test.go:385:      TotalAlloc = 5 MiB
    task_pool_test.go:386:      Sys = 69 MiB
    task_pool_test.go:387:      NumGC = 2
    task_pool_test.go:393: 
                 TestWithPoolUseAlways took 65.361076ms
PASS
```

**Review by fangyincheng — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Pull request merged**